metals/src/main/scala/scala/meta/internal/metals/debug/DebugServer.scala

```diff
@@ -105,6 +112,155 @@ object DebugServer {
+              ): Try[(b.ScalaMainClass, b.BuildTargetIdentifier)]
+            } { clazz => Success(clazz -> target.getId()) }
+        }
+    }
+    classAndTarget.map {
+      case (clazz, target) =>
+        clazz.setArguments(Option(params.args).getOrElse(List().asJava))
+        clazz.setJvmOptions(Option(params.jvmOptions).getOrElse(List().asJava))
+        val dataKind =
+          b.DebugSessionParamsDataKind.SCALA_MAIN_CLASS
+        val data = clazz.toJson
+        new b.DebugSessionParams(
+          singletonList(target),
+          dataKind,
+          data
+        )
+    }
+  }
+
+  def resolveTestClassParams(
+      params: DebugUnresolvedTestClassParams,
+      buildTargets: BuildTargets,
+      buildTargetClasses: BuildTargetClasses,
+      showWarningMessage: String => Unit
+  ): Try[b.DebugSessionParams] = {
+    val classAndTarget = Option(params.buildTarget).fold {
+      val classes =
+        buildTargetClasses.findTestClassByName(params.testClass)
+      if (classes.nonEmpty) Success {
+        val classAndTarget = classes.head
+        if (classes.length > 1) {
+          val targetName =
+            buildTargets.info(classAndTarget._2).get.getDisplayName
+          val anotherTargets = classes.tail
+            .map {
+              case (_, id) =>
+                buildTargets.info(id).get.getDisplayName
+            }
+          showWarningMessage(
+            UnresolvedDebugSessionParams
+              .runningClassMultipleBuildTargetsMessage(
+                classAndTarget._1,
+                targetName,
+                anotherTargets,
+                "test"
+              )
+          )
+        }
+        classAndTarget
+      }
+      else
+        Failure(new ClassNotFoundException(params.testClass))
+    } { targetName =>
+      buildTargets
+        .findByDisplayName(targetName)
+        .fold {
+          Failure[(String, b.BuildTargetIdentifier)](
+            new BuildTargetNotFoundException(targetName)
+          ): Try[(String, b.BuildTargetIdentifier)]
+        } { target =>
+          buildTargetClasses
+            .classesOf(target.getId())
+            .testClasses
+            .values
+            .find(
+              _ == params.testClass
+            )
+            .fold {
+              Failure[(String, b.BuildTargetIdentifier)](
+                new ClassNotFoundInBuildTargetException(
+                  params.testClass,
+                  targetName
+                )
+              ): Try[(String, b.BuildTargetIdentifier)]
+            } { clazz =>
+              Success(
+                clazz -> target.getId()
+              )
+            }
+        }
+    }
+    classAndTarget.map {
+      case (clazz, target) =>
+        val dataKind =
+          b.DebugSessionParamsDataKind.SCALA_TEST_SUITES
+        val data = singletonList(clazz).toJson
+        new b.DebugSessionParams(
+          singletonList(target),
+          dataKind,
+          data
+        )
+    }
+  }
```

Comment:
Could you split the method somehow?

Reply:
> I've moved finding class and build target data into new class BuildTargetClassesFinder.
findMainClassAndItsBuildTarget and findTestClassAndItsBuildTarget can also probably can be refactored, especially since they follow the same scenario, just main/test is different, but I'm a bit afraid, it eventually get complex-easy (instead of a simple-hard).